File: .algorithm_and_structure/unionfind.c
#include <assert.h>
#include <stdlib.h>

void swap(int *x, int *y) {
  int tmp;
  tmp = *x;
  *x = *y;
  *y = tmp;
}

typedef struct {
  int _n;
  int *parent;
} UnionFind;

UnionFind make_unionfind(int n) {
  UnionFind uf;
  uf._n = n;
  uf.parent = (int *)malloc(sizeof(int) * n);
  for (int i = 0; i < n; i++) {
    uf.parent[i] = -1;
  }
  return uf;
}

void delete_unionfind(UnionFind *ufp) { free(ufp->parent); }

int find(UnionFind *ufp, int i) {
  assert(0 <= i && i < ufp->_n);
  if (ufp->parent[i] < 0) {
    return i;
  } else {
    return ufp->parent[i] = find(ufp, ufp->parent[i]);
  }
}

int merge(UnionFind *ufp, int i, int j) {
  assert(0 <= i && i < ufp->_n);
  assert(0 <= j && j < ufp->_n);
  int x, y;
  x = find(ufp, i);
  y = find(ufp, j);
  if (x == y) {
    return x;
  }
  if (-ufp->parent[x] < -ufp->parent[y]) {
    swap(&x, &y);
  }
  ufp->parent[x] += ufp->parent[y];
  return ufp->parent[y] = x;
}

int size(UnionFind uf) { return uf._n; }

int size_group(UnionFind *ufp, int i) {
  assert(0 <= i && i < ufp->_n);
  return -ufp->parent[find(ufp, i)];
}

int count_group(UnionFind uf) {
  int result = 0;
  for (int i = 0; i < uf._n; i++) {
    if (uf.parent[i] < 0) {
      result++;
    }
  }
  return result;
}

File: .algorithm_and_structure/unionfind.py
class UnionFind():
  def __init__(self, n):
    self.__size = n
    self.__parent = [-1] * n

  def find(self, i):
    assert 0 <= i < self.__size
    if self.__parent[i] < 0:
      return i
    else:
      self.__parent[i] = self.find(self.__parent[i])
      return self.__parent[i]

  def union(self, i, j):
    assert 0 <= i < self.__size
    assert 0 <= j < self.__size
    x,y = self.find(i),self.find(j)
    if x == y:
      return x
    elif -self.__parent[x] < -self.__parent[y]:
      x,y = y,x
    self.__parent[x] += self.__parent[y]
    self.__parent[y] = x
    return x

  def same(self, i, j):
    assert 0 <= i < self.__size
    assert 0 <= j < self.__size
    return self.find(i) == self.find(j)

  def size(self, i):
    assert 0 <= i < self.__size
    return -self.__parent[self.find(i)]

  def roots(self):
    return [i for i,x in enumerate(self.__parent) if x < 0]

  def count_group(self):
    return len(self.roots())
